File: src/module/zing-captcha-solver.ts
import { join } from 'node:path';
import * as ort from 'onnxruntime-node';
import sharp from 'sharp';

const getElectronApp = (): { isPackaged: boolean } | undefined => {
    try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const electron = require('electron');
        return electron.app;
    } catch {
        return undefined;
    }
};

class ZingCaptchaSolver {
    private static readonly CHAR_SET = ['3', '6', '7', '8', '9', 'B', 'D', 'E', 'F', 'H', 'K', 'M', 'N', 'P', 'R', 'T', 'U', 'V', 'W', 'X', 'Y'];
    private static readonly BLANK_LABEL = ZingCaptchaSolver.CHAR_SET.length;
    private static readonly IDX_TO_CHAR = Object.fromEntries(ZingCaptchaSolver.CHAR_SET.map((c, i) => [i, c]));

    private session: ort.InferenceSession | null = null;
    private modelPath: string | null = null;

    private readonly getModelPath = (): string => {
        if (this.modelPath) return this.modelPath;
        const app = getElectronApp();
        this.modelPath = app?.isPackaged ? join(process.resourcesPath, 'models/zing-captcha-crnn-ocr.onnx') : join(process.cwd(), 'models/zing-captcha-crnn-ocr.onnx');
        return this.modelPath;
    };

    async solve(url: string): Promise<string> {
        this.session ??= await ort.InferenceSession.create(this.getModelPath());

        const response = await fetch(url);
        if (!response.ok) throw new Error(`lỗi fetch: ${response.status}`);

        const buffer = Buffer.from(await response.arrayBuffer());
        const { data, info } = await sharp(buffer).resize(200, 50).grayscale().raw().toBuffer({ resolveWithObject: true });

        const normalized = new Float32Array(data.length);
        for (let i = 0; i < data.length; i++) {
            normalized[i] = ((data[i] ?? 0) / 255 - 0.5) / 0.5;
        }

        const inputTensor = new ort.Tensor('float32', normalized, [1, 1, info.height ?? 50, info.width ?? 200]);

        const outputs = await this.session.run({ input: inputTensor });
        const outputTensor = outputs['output'];
        if (!outputTensor?.data || !outputTensor?.dims) {
            throw new Error('lỗi output model');
        }

        const output = outputTensor.data as Float32Array;
        const shape = outputTensor.dims;
        const seqLen = shape[0] ?? 0;
        const numClasses = shape[2] ?? 0;

        const predIdx: number[] = [];
        for (let t = 0; t < seqLen; t++) {
            let maxIdx = 0;
            let maxVal = -Infinity;
            for (let c = 0; c < numClasses; c++) {
                const val = output[t * numClasses + c] ?? -Infinity;
                if (val > maxVal) {
                    maxVal = val;
                    maxIdx = c;
                }
            }
            predIdx.push(maxIdx);
        }

        const result: string[] = [];
        let prev: number | null = null;
        for (const idx of predIdx) {
            if (idx !== ZingCaptchaSolver.BLANK_LABEL && idx !== prev) {
                const char = ZingCaptchaSolver.IDX_TO_CHAR[idx];
                if (char) result.push(char);
            }
            prev = idx;
        }

        return result.join('');
    }
}
export default ZingCaptchaSolver;

File: utils/zing_captcha_solver.py
from pathlib import Path
import io
import sys
import os
from typing import Optional
import onnxruntime as ort
from PIL import Image
import numpy as np
import requests
import traceback


def get_base_path():
    if getattr(sys, "frozen", False):
        exe_path = os.path.abspath(sys.executable)
        return Path(os.path.dirname(exe_path))
    return Path(__file__).parent.parent


CHAR_SET = [
    "3",
    "6",
    "7",
    "8",
    "9",
    "B",
    "D",
    "E",
    "F",
    "H",
    "K",
    "M",
    "N",
    "P",
    "R",
    "T",
    "U",
    "V",
    "W",
    "X",
    "Y",
]
BLANK_LABEL = len(CHAR_SET)
IDX_TO_CHAR = dict(enumerate(CHAR_SET))


class ZingCaptchaSolver:
    def __init__(self, model_path=None):
        if model_path is None:
            base_path = get_base_path()
            model_path = base_path / "models" / "zing-captcha-crnn-ocr.onnx"
        self.model_path = Path(model_path)
        self.session: Optional[ort.InferenceSession] = None

    def _load_session(self) -> None:
        if self.session is None:
            self.session = ort.InferenceSession(str(self.model_path))

    def _preprocess(self, img_data: bytes) -> np.ndarray:
        img = Image.open(io.BytesIO(img_data)).convert("L")
        img = img.resize((200, 50))
        img_arr = np.array(img).astype(np.float32) / 255.0
        img_arr = (img_arr - 0.5) / 0.5
        img_arr = img_arr[np.newaxis, np.newaxis, :, :]
        return img_arr

    def _ctc_decode(self, pred_idx: np.ndarray) -> str:
        result = []
        prev = None
        for idx in pred_idx:
            if idx != BLANK_LABEL and idx != prev:
                result.append(IDX_TO_CHAR[idx])
            prev = idx
        return "".join(result)

    def solve(self, url: str) -> str:
        try:
            self._load_session()

            response = requests.get(url, timeout=10)
            response.raise_for_status()

            input_tensor = self._preprocess(response.content)

            if self.session is None:
                raise RuntimeError("session chưa được load")
            outputs = self.session.run(None, {"input": input_tensor})
            output = np.array(outputs[0])

            pred_idx = output.argmax(axis=2).squeeze()

            result = self._ctc_decode(pred_idx)
            return result
        except Exception:
            traceback.print_exc()
            raise
